perf/inferenceql/query/perf.clj

```diff
@@ -0,0 +1,142 @@
+(ns inferenceql.query.perf
+  (:require [clj-async-profiler.core :as prof]
+            [clojure.java.browse :as browse]
+            [clojure.string :as str]
+            [criterium.core :as crit]
+            [inferenceql.query.strict :as strict]
+            [medley.core :as medley]))
+
+(def ^:dynamic *default-limit* 10000)
+
+(defn- str-listify
+  [x]
+  (if (string? x) x (str/join ", " x)))
+
+(defn conditioned-prob-density
+  [{:keys [model table prob-density-evt conditioned-density-evt]}]
+  (str "SELECT PROBABILITY DENSITY OF " prob-density-evt
```

Comment:
Any concern about SQL injection attack? Is there a SQL language escape function handy with which we could wrap the things we're going to interpolate into SQL language strings here?

Reply:
I agree it doesn't have to happen for this PR (*) but what kind of code reviewer would I be if I didn't show respect to Bobby Tables. 

(*) or maybe ever, since this is just a way of taking some query material and wrapping it in a family of benchmark scenarios. I haven't done SQL + clojure before, so I was figuring there must be some way of safely assembling SQL statements that might be used elsewhere in the codebase (someday, it will be an issue)

Reply:
Heh. "And I hope _you've_ learned to sanitize your database inputs."

There's a wide variety of SQL libs on https://www.clojure-toolbox.com/. (I've used HoneySQL and HugSQL in the past, but I'm not the biggest fan. They're safe, though.) However, I think most SQL injection prevention happens a little lower down, at the JDBC driver layer.

Parameterized statements are more common than escaping, at least in the Java world. Partly because you can combine them with _prepared_ statements, for efficiency (like a saved execution plan), and partly because many databases support binary transfer of parameters, so there's no need to convert to/from a string.

In next.jdbc, the de facto clj standard, it looks something like `["SELECT * FROM foo WHERE x = ? AND y = ?" my-x-value my-y-value]`, where you supply a query with ?'s for params, followed by the values you want to send. Underneath the hood, the database's JDBC driver will ensure it's all safe.

Problem is, that makes us the equivalent of the JDBC driver authors. We could do something similar, leave placeholders in the syntax tree, and then drop in the str versions of whatever was passed-in, for compatibility. That way it could never parse maliciously.